src/toxcore/stats.rs

```diff
@@ -0,0 +1,110 @@
+/*!
+Statistics of incoming/outgoing packets
+This is used by both Udp codec and Tcp codec.
+*/
+
+#[cfg(target_pointer_width = "64")]
+use std::sync::atomic::*;
+#[cfg(target_pointer_width = "32")]
+use std::sync::Mutex;
+use std::sync::Arc;
+
+#[derive(Clone, Debug, Default)]
+/// Struct for counting packets on x64 CPU.
+#[cfg(target_pointer_width = "64")]
+pub struct Stats {
+    /// Incoming packets count for Udp/Tcp
+    incoming: Arc<AtomicUsize>,
+    /// Outgoing packets count for Udp/Tcp
+    outgoing: Arc<AtomicUsize>,
+}
+
+#[derive(Clone, Debug, Default)]
+/// Struct for counting packets on x86 CPU.
+#[cfg(target_pointer_width = "32")]
+pub struct Stats {
+    /// Incoming packets count for Udp/Tcp
+    pub incoming: Arc<Mutex<u64>>,
+    /// Outgoing packets count for Udp/Tcp
+    pub outgoing: Arc<Mutex<u64>>,
+}
+
+impl Stats {
+    /// New Stats object
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    /// Add 1 to incoming counter
+    #[cfg(target_pointer_width = "64")]
+    pub fn increase_incoming(&self) {
+        self.incoming.fetch_add(1, Ordering::SeqCst);
+    }
+
+    /// Add 1 to incoming counter
+    #[cfg(target_pointer_width = "32")]
+    pub fn increase_incoming(&self) {
+        *self.incoming.lock().expect("Can't lock mutex") += 1;
+    }
+
+    /// Add 1 to incoming counter
+    #[cfg(target_pointer_width = "64")]
+    pub fn increase_outgoing(&self) {
+        self.outgoing.fetch_add(1, Ordering::SeqCst);
+    }
+
+    /// Add 1 to incoming counter
+    #[cfg(target_pointer_width = "32")]
+    pub fn increase_outgoing(&self) {
+        *self.outgoing.lock().expect("Can't lock mutex") += 1;
+    }
+
+    /// Get incoming count
+    #[cfg(target_pointer_width = "64")]
+    pub fn get_current_incoming(&self) -> u64 {
+        self.incoming.load(Ordering::SeqCst) as u64
+    }
+
+    /// Get incoming count
+    #[cfg(target_pointer_width = "32")]
+    pub fn get_current_incoming(&self) -> u64 {
+        *self.incoming.lock().expect("Can't lock mutex")
+    }
+
+    /// Get outgoing count
+    #[cfg(target_pointer_width = "64")]
+    pub fn get_current_outgoing(&self) -> u64 {
+        self.outgoing.load(Ordering::SeqCst) as u64
+    }
+
+    /// Get outgoing count
+    #[cfg(target_pointer_width = "32")]
+    pub fn get_current_outgoing(&self) -> u64 {
+        *self.outgoing.lock().expect("Can't lock mutex")
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn new() {
+        let _stats = Stats::new();
+    }
+
+    #[test]
+    fn get_current_incoming() {
+        let stats = Stats::new();
+        stats.increase_incoming();
```

Comment:
Test that stats are initialized with 0

Reply:
@NamsooCho the last notes

Reply:
https://github.com/tox-rs/tox/blob/d3259861fdecd452836fadd1d2f756648ecaad8a/src/toxcore/stats.rs#L107
It may be redundant.